src/Validation/Validator.php

```diff
@@ -813,95 +879,221 @@ public function allowEmptyString($field, $when = true, $message = null)
      * EMPTY_ARRAY flags.
      *
      * @param string $field The name of the field.
+     * @param string|null $message The message to show if the field is not
      * @param bool|string|callable $when Indicates when the field is allowed to be empty
      * Valid values are true, false, 'create', 'update'. If a callable is passed then
      * the field will allowed to be empty only when the callback returns true.
-     * @param string|null $message The message to show if the field is not
      * @return $this
      * @since 3.7.0
-     * @see \Cake\Validation\Validator::allowEmptyFor() For detail usage
+     * @see \Cake\Validation\Validator::allowEmptyFor() for examples.
      */
-    public function allowEmptyArray($field, $when = true, $message = null)
+    public function allowEmptyArray($field, $message = null, $when = true)
     {
+        list($message, $when) = $this->sortMessageAndWhen($message, $when, __METHOD__);
+
+        return $this->allowEmptyFor($field, self::EMPTY_STRING | self::EMPTY_ARRAY, $when, $message);
+    }
+
+    /**
+     * Require a field to be a non-empty array
+     *
+     * Opposite to allowEmptyArray()
+     *
+     * @param string $field The name of the field.
+     * @param string|null $message The message to show if the field is empty.
+     * @param bool|string|callable $when Indicates when the field is not allowed
+     *   to be empty. Valid values are false (never), 'create', 'update'. If a
+     *   callable is passed then the field will be required to be not empty when
+     *   the callback returns true.
+     * @return $this
+     * @see \Cake\Validation\Validator::allowEmptyArray()
+     * @since 3.8.0
+     */
+    public function notEmptyArray($field, $message = null, $when = false)
```

Comment:
could we have `$when = true`  by default ?
now callback can be return `true` ( as `false` after calling `invertWhenClause()`) ,
and this can be make confusing 
```php
        } elseif (is_callable($when)) {
            return function ($context) use ($when) {
                return !$when($context);
            };
        }
```
my suggest : 
`create` : can not be empty `when==create` // `update` is allowed
`update` : can not be empty `when==update` // `create` is allowed
`true` : can not be empty in both `create` + `update` (never)  
`false` : can be empty in both `create` + `update` // negates the behavior like `allowEmpty*()`
with this change return of `callback` and and method argument will same ( `true` as is not allowed )

this change can be hard to refactoring from `notEmpty()` to `notEmpty*()`  methods

Reply:
The condition inversion was part of how notEmpty() worked as well. That is why I made these methods default to false as well.

The inversion is required because when rules are evaluated, the validator needs to know if a field can be empty or not  as these methods and the allowEmpty* methods change the same internal state.

Reply:
Please do not comment on old PRs, open a new ticket or ask in the public channel if you have a question about future changes in the core.